train_gpt2.cu

```diff
@@ -256,6 +260,29 @@ __device__ floatX warpReduceSum(floatX val) {
 }
 #endif
 
+using reduction_func_t = float (*) (float);
```

Comment:
not really a fan of the function pointer. Can we make this a template with a (compile-time constant) function pointer argument?

Reply:
NASA approves of removing this i think, can be done later

Reply:
```c++
template<reduction_func_t warp_reduction>
__device__ float blockReduce(float val, bool final_sync=false)
```
?
Haven't tried this, but it's how the kernel launcher works in one of the dev/cuda files

Reply:
Yep that worked, I changed it to that instead of a function pointer for now (agreed with your other comment we could consider changing it a functor in the future).